### PR TITLE
Set Cartesian axis attributes via template

### PR DESCRIPTION
**Pull request by archmoj**
Commit https://github.com/plotly/plotly.js/pull/4670/commits/5f8b280fc78b0dd8ad88440a317cb0d35811bb24 addresses #4667 i.e. to set axis `type`, `rangebreaks` and `tickformatstops` from 
`template`.

@plotly/plotly_js

### CODE REVIEW

**Comment by nicolaskruchten** [comment]
I think we should leave #4671 alone for now please... this code already works as desired for e.g. annotations. Check out this Python example: https://plot.ly/python/templates/#including-tuple-elements-in-a-theme

**Comment by archmoj** [comment]
In JS, it appears those were not working [before](https://codepen.io/MojtabaSamimi/pen/xxGaMbv?editors=0010) | [after](https://codepen.io/MojtabaSamimi/pen/qBdMgZE?editors=1000).

@nicolaskruchten @alexcjohnson am I missing something?

**Comment by nicolaskruchten** [comment]
This works for me with the production bundle: https://codepen.io/nicolaskruchten/pen/wvaENrm?editors=0010

**Comment by nicolaskruchten** [comment]
@alexcjohnson I'll let you judge the safety of these changes, keeping in mind that I would rather keep existing bugs in the interests of getting 1.53 and 1.54 done ASAP and not introducing subtle new bugs. For 1.53 we just need `rangebreaks` and `hovermode = "* unified"` to work with templates IMO.

**Comment by alexcjohnson** [comment]
@archmoj the way templates interact with container arrays is a bit subtle as it contains features for both defaults for all array items and overrides for specific items (which may not be at the same index in the array), but in general it works. Take a look at the original templates PR https://github.com/plotly/plotly.js/pull/2761, esp. [this test](https://github.com/plotly/plotly.js/pull/2761/files#diff-10fdc9e067ffc2e976a05e6d98cebbf1R57) and the [`arrayTemplater` function](https://github.com/plotly/plotly.js/pull/2761/files#diff-44b7021288b6b3409a7d11ef712b75eeR144)

Please use those for context before deciding what's a bug.

**Comment by nicolaskruchten** [comment]
Ah, with respect to your before/after: you need to set `name` for that to work... https://plot.ly/javascript/reference/#layout-annotations-items-annotation-name

File: src/plots/cartesian/layout_defaults.js [hunk 1]
```diff
@@ -230,7 +230,13 @@ module.exports = function supplyLayoutDefaults(layoutIn, layoutOut, fullData) {
                 axHasImage[axName]
               ));
 
+        var axTemplate;
+        if(Lib.isPlainObject(layoutOut._template)) {
+            axTemplate = layoutOut._template[axLayoutOut._name];
```
Comment (alexcjohnson) [inline src/plots/cartesian/layout_defaults.js]: This isn't quite right - `Template.newContainer` also lets `xaxis` from the template be used for `xaxis2`, xaxis3` etc in the actual plot. But do you need `axTemplate` at all? Can't you just reference `containerOut._template` where you currently reference `options.axTemplate`?

Comment (archmoj) [reply src/plots/cartesian/layout_defaults.js]: @alexcjohnson Good call. Fixed in https://github.com/plotly/plotly.js/pull/4670/commits/2494c4e1ad8718bcd4e23d559e3caa0509162a44.

File: src/plots/cartesian/axis_defaults.js [hunk 2]
```diff
@@ -124,8 +125,9 @@ module.exports = function handleAxisDefaults(containerIn, containerOut, coerce,
         }
     }
 
-    if(containerOut.type === 'date') {
-        var rangebreaks = containerIn.rangebreaks;
+    if(axType === 'date') {
+        var rangebreaks = containerIn.rangebreaks || axTemplate.rangebreaks;
```
Comment (alexcjohnson) [inline src/plots/cartesian/axis_defaults.js]: This isn't quite right... `[]` is truthy, also `axTemplate.rangeBreaks` items will extend the array if it contains named items.

Can we instead do:

```js
handleArrayContainerDefaults(...);

if (containerOut.rangebreaks.length) {
   // setConvert and gl / splom check
}
else {
   delete containerOut.rangebreaks;
}
```

Same for `tickformatstops`

Comment (archmoj) [reply src/plots/cartesian/axis_defaults.js]: Good idea. Done in https://github.com/plotly/plotly.js/pull/4670/commits/f824902d0c044d3fcd7342b45efd20461248f6d1.

**Pull request merged**